### PR TITLE
Re-land: emit macro expansion stack in diagnostics, with O(1)-per-invocation tracking

### PR DESCRIPTION
**Pull request by zangold-nv**
Motivation

  PR #6165 added "expanded from macro 'X'" notes to Slang diagnostics so users can trace errors back through
  macro call sites. That PR was subsequently reverted in #11116 due to a performance regression in core-module
  compilation.

  The regression came from the implementation strategy: for every macro invocation with a non-empty body, the
  old code called createSourceFileWithBlob + createSourceView, which:

  - Allocated a full-file-sized chunk of the global 32-bit source-location namespace (up to 1.2 MB per
  invocation for hlsl.meta.slang)
  - Heap-allocated a SourceFile and SourceView object, appending to m_sourceViews
  - Made every subsequent findSourceView binary search slower as m_sourceViews grew

  This PR re-lands the feature with a new implementation that eliminates all three costs.

  Proposed solution

  Replace the per-invocation SourceFile/SourceView approach with a flat side table (List<MacroExpansionEntry> on
   SourceManager). Each entry stores the macro name, call-site loc, first body token loc, and a tiny allocated
  source range (bodyEnd - bodyBegin + 1 bytes — a few dozen bytes, not 1.2 MB). No SourceFile or SourceView is
  created; m_sourceViews does not grow.

  Because the tiny expansion ranges have no SourceView, three callsites that previously relied on findSourceView
   to resolve diagnostic locs now call a new findSourceViewThroughExpansion helper that walks the side table to
  unmap a remapped loc back to its original definition-file loc. This preserves source context (underlines, span
   messages, line/column info) in both the text and rich diagnostic renderers.

  The diagnostic chain walkers (formatDiagnostic and appendMacroExpansionNotes) query findMacroExpansion() for
  macro-expansion levels and continue to use the existing SourceView initiating-loc chain for TokenPaste levels,
   so both features work correctly together.

  Change summary

  File: slang-source-loc.h/cpp
  What changed: Added MacroExpansionEntry, registerMacroExpansion, findMacroExpansion,
    findSourceViewThroughExpansion; updated getHumaneLoc to use the new helper; added m_macroExpansions.clear()
    to _resetSource; removed PathInfo::Type::MacroExpansion
  ────────────────────────────────────────
  File: slang-preprocessor.cpp
  What changed: MacroInvocation constructor calls registerMacroExpansion instead of
    createSourceFileWithBlob+createSourceView; _remapBodyLoc uses (loc − bodyBegin) offset into the tiny range
  ────────────────────────────────────────
  File: slang-diagnostic-sink.cpp
  What changed: Both the text-path loop (formatDiagnostic) and the rich-path helper (appendMacroExpansionNotes)
    query the side table for macro-expansion levels; use findSourceViewThroughExpansion to resolve locs for
    humane loc display
  ────────────────────────────────────────
  File: slang-rich-diagnostics-render.cpp
  What changed: Two findSourceView calls replaced with findSourceViewThroughExpansion so source context renders
    correctly for macro-body tokens
  ────────────────────────────────────────
  File: slang-compile-request.cpp
  What changed: Removed PathInfo::Type::MacroExpansion case (type no longer exists)
  ────────────────────────────────────────
  File: tests/diagnostics/macro-expansion-stack/
  What changed: Six tests covering single-level, nested, diamond, function-like, argument, and empty macros
  ────────────────────────────────────────
  File: tests/diagnostics/token-paste-location.slang
  What changed: Test covering interleaved macro-expansion and token-paste chains

  Concepts and vocabulary

  Expansion range: The tiny SourceRange allocated per invocation, covering bodyEnd − bodyBegin + 1 raw source
  IDs. Body token locs are remapped into this range so tokens from different invocations of the same macro are
  distinguishable. No SourceFile or SourceView backs this range.

  Unmapping: Given a remapped loc in an expansion range, recovering the original definition-file loc via
  bodyBegin + (remappedLoc − range.begin). Used in findSourceViewThroughExpansion and the diagnostic chain
  walkers so that source context (line content, column, span messages) can be resolved.

  MacroExpansionEntry: One record in the side table — the macro name (for the note), the call-site loc (for
  walking up the expansion chain), the body-begin loc (for unmapping), and the allocated range.

  Process report

  Original regression cause. The old implementation called allocateSourceRange(blob->getBufferSize()) per
  invocation. For hlsl.meta.slang (1.2 MB), 153+ invocations of the six macros defined there consumed ~180
  million of the 4-billion available 32-bit source IDs and grew m_sourceViews by 153+ entries, slowing every
  subsequent binary search in that list.

  New allocation. allocateSourceRange is now called with bodyEnd.getRaw() − bodyBegin.getRaw() + 1. For a
  typical macro body this is tens to hundreds of bytes. The side table entry is a plain struct (SourceRange +
  two SourceLocs + a String); no heap indirection beyond the string.

  Why findSourceViewThroughExpansion rather than modifying findSourceView. findSourceView is a hot path called
  from many places. Adding a side-table lookup there would add overhead on every call, even when the loc is in a
   normal SourceView. The new helper is called only from the three places that need to display source context
  for a loc that may be in an expansion range: getHumaneLoc, and two spots in the rich renderer.

  Chain walking for mixed macro/token-paste stacks. Body token locs (including the ## operator's loc) are
  remapped into the expansion range, so the TokenPaste SourceView's initiatingSourceLoc may itself be a remapped
   loc with no backing SourceView. The chain walkers therefore try the macro-expansion side table first; if the
  current loc is in a macro expansion range they emit the "expanded from" note and follow callSiteLoc. If not,
  they look for a TokenPaste SourceView and, when resolving the paste's initiatingSourceLoc for display, call
  findSourceViewThroughExpansion to unmap any nested remapped loc. This preserves the full expansion stack
  across interleaved macro and paste levels.

### CODE REVIEW

File: source/compiler-core/slang-diagnostic-sink.cpp [hunk 1]
```diff
@@ -433,17 +433,25 @@ static void _tokenLengthNoteDiagnostic(
     }
 }
 
+/// Find the SourceView for loc, walking through macro expansion side-table entries when
+/// findSourceViewRecursively returns null (because loc is in a remapped expansion range with no
+/// SourceView). Updates loc in-place to the original (un-remapped) loc on success.
```
Comment (tangent-vector) [inline source/compiler-core/slang-diagnostic-sink.cpp]: Please reframe this documentation comment so that it is entirely in terms of the client-facing behavior and semantic contract of the operation, rather than describing the operation in terms of implementation details (e.g., "when `findSourceViewRecursively` returns `null`").

File: source/compiler-core/slang-diagnostic-sink.cpp [hunk 2]
```diff
@@ -433,17 +433,25 @@ static void _tokenLengthNoteDiagnostic(
...
+static SourceView* findSourceViewThroughMacroExpansion(SourceManager* sm, SourceLoc& loc)
+{
+    return sm ? sm->findSourceViewThroughExpansion(loc) : nullptr;
```
Comment (tangent-vector) [inline source/compiler-core/slang-diagnostic-sink.cpp]: This is another case of a cowardly/CYA conditional on the part of an agent. As is the general rule, this early-out / null check should either:

- Actually be an assertion of an explicit precondition of the operation (`sm` must not be null)... at which point I'm not sure what point this helper routine serves...

- Include a clear rationale for (1) why the condition is allowed/expected to occur (why must `sm` be null along some code paths?), and (2) a rock-solid explanation/proof for why the handling of that condition being performed here is the clearly Correct one.

Also, if `sm` is allowed to be null, then the documentation comment for the subroutine should state this, the explanation for why, and a description of what the semantics of the operation are that make that case valid and a match for the name given to this subroutine. As it is this subroutine should seemingly be named `findSourceViewThroughMacroExpansionIfSourceManagerIsNotNull`... which is clearly a silly operation to define, and would be a code smell if this were named honestly...

File: source/compiler-core/slang-diagnostic-sink.cpp [hunk 3]
```diff
@@ -433,17 +433,25 @@ static void _tokenLengthNoteDiagnostic(
...
 
     SourceView* sourceView = nullptr;
     HumaneSourceLoc humaneLoc;
-    const auto sourceLoc = diagnostic.loc;
+    auto sourceLoc = diagnostic.loc;
     {
         if (sourceManager)
         {
-            sourceView = sourceManager->findSourceViewRecursively(sourceLoc);
+            sourceView = findSourceViewThroughMacroExpansion(sourceManager, sourceLoc);
```
Comment (tangent-vector) [inline source/compiler-core/slang-diagnostic-sink.cpp]: Based on this use site, it seems like the operation should be named in a way that makes the intent clear: this seems to be an operation that returns a source view corresponding to the source location that includes the chain of expansion (macro expansion, token pastes, etc.), whereas (I presume) other queries for a source view would only provide the "spelling" location (?).

Maybe I should be looking at the header first, before this code, but it seems like a case where the naming conventions should be making the intent much more clear to the reader. As it is, a reviewer (like me) could reasonably ask: why is finding a source view "through macro expansion" the correct thing here, as opposed to finding the source view "recursively"? What is the difference between these operations and what they mean/return?

File: source/compiler-core/slang-diagnostic-sink.cpp [hunk 4]
```diff
@@ -453,44 +461,87 @@ static void formatDiagnostic(DiagnosticSink* sink, Diagnostic const& diagnostic,
...
         {
-            SourceView* currentView = sourceView;
-
-            while (currentView && currentView->getInitiatingSourceLoc().isValid() &&
-                   currentView->getSourceFile()->getPathInfo().type == PathInfo::Type::TokenPaste)
+            // Walk the mixed macro-expansion/token-paste chain starting from the diagnostic loc.
+            // MacroExpansion levels are found via the SourceManager side table (no SourceViews);
+            // TokenPaste levels are found via the existing SourceView initiating-loc chain.
+            SourceLoc currentLoc = diagnostic.loc;
+            for (int expansionDepth = 0; expansionDepth < 1024; ++expansionDepth)
```
Comment (tangent-vector) [inline source/compiler-core/slang-diagnostic-sink.cpp]: Please turn the magic number into a named constant.

File: source/compiler-core/slang-diagnostic-sink.cpp [hunk 5]
```diff
@@ -453,44 +461,87 @@ static void formatDiagnostic(DiagnosticSink* sink, Diagnostic const& diagnostic,
...
+                    initiationDiagnostic.ErrorID = MiscDiagnostics::seeExpandedFromMacro.id;
+                    initiationDiagnostic.Message = msg.produceString();
+                    initiationDiagnostic.loc = callSiteLoc;
+                    initiationDiagnostic.severity = MiscDiagnostics::seeExpandedFromMacro.severity;
+
+                    formatDiagnostic(
+                        callSiteView->getHumaneLoc(callSiteLoc),
+                        initiationDiagnostic,
+                        sink->getFlags(),
+                        sb);
```
Comment (tangent-vector) [inline source/compiler-core/slang-diagnostic-sink.cpp]: I initially got worried upon seeing a call to `formatDiagnostic` from deep within a loop inside `formatDiagnostic` itself; this seems like a recipe for an infinite recursion. Upon closer inspection the signatures are different, though, so *hopefully* this isn't actually a potential recursion path.

It seems like realistically there are multiple distinct operations all going under the name `formatDiagnostic` which have meaningfully distinct semantics that probably shouldn't be obscured by giving them all the same simple/pithy name. The one we're looking at here seems to not just format a single diagnostic but can, in fact, format a whole "chain" of diagnostics include their expansion provenance chain, but the function name doesn't make that part clear.

I would like to see the routines named in a more complete/logical fashion to make the layering more clear, and remove the fear of infinite recursion that a call like this creates for a reader just browsing the code.

File: source/compiler-core/slang-diagnostic-sink.cpp [hunk 6]
```diff
@@ -453,44 +461,87 @@ static void formatDiagnostic(DiagnosticSink* sink, Diagnostic const& diagnostic,
...
-                // Make the initiatingView the current view
-                currentView = initiatingView;
+                initiationDiagnostic.loc = initiatingLoc;
+                initiationDiagnostic.severity = MiscDiagnostics::seeTokenPasteLocation.severity;
+
+                formatDiagnostic(
+                    initiatingView->getHumaneLoc(initiatingLoc),
+                    initiationDiagnostic,
+                    sink->getFlags(),
+                    sb);
```
Comment (tangent-vector) [inline source/compiler-core/slang-diagnostic-sink.cpp]: It seems undesirable (a bit of a code smell) that the way this loop body is currently written here has a duplication of the logic that actually formats the `initiatingDiagnostic`. It's not immediately clear to me what the control flow has this odd one-sided `if` with a `continue`, that duplications code with the not-taken path, rather than a plain old two-sided `if`/`else` to handle the code that differs based on whether or not we find a `macroEntry`, before bottlenecking down to the shared code that runs in either case.

A big part of the problem is the lack of actual comments explaining the solution approach being taken here, and walking a reader through the flow of the code and why it is the way it is (not just comments saying what the next line of code does, but comments that explain the actual algorithm/approach the author had in mind).

File: source/compiler-core/slang-diagnostic-sink.cpp [hunk 7]
```diff
@@ -632,6 +683,69 @@ bool DiagnosticSink::diagnoseRichImpl(
...
 }
 
+/// Walk the mixed macro-expansion/token-paste chain rooted at `primaryLoc` and append a note for
+/// each level. MacroExpansion levels are found via the SourceManager side table; TokenPaste levels
+/// are found via the existing SourceView initiating-loc chain.
+static void appendMacroExpansionNotes(
+    SourceManager* sm,
+    SourceLoc primaryLoc,
+    List<DiagnosticNote>& notes)
+{
```
Comment (tangent-vector) [inline source/compiler-core/slang-diagnostic-sink.cpp]: How is what this routine does different from what was in the other loop above, and what is the justification for the high degree of code duplication between them? It really seems like there's a set of shared helper routines begging to be defined here, with more of the code just bottlenecking through a few small helpers (actually `appendMacroExpansionNotes` might be one of the helpers that is desirable, and its just that the code above is failing to *use* it).

At least this subroutine has a nice and clear name that actually seems to match what the subroutine *does*.

File: source/compiler-core/slang-diagnostic-sink.cpp [hunk 8]
```diff
@@ -632,6 +683,69 @@ bool DiagnosticSink::diagnoseRichImpl(
...
+    if (!sm || !primaryLoc.isValid())
+        return;
```
Comment (tangent-vector) [inline source/compiler-core/slang-diagnostic-sink.cpp]: More cowardly/CYA coding. The broken record is that this should either:

- Assert its preconditions rather than defensively test them, like a grown-up.
- Include a clear rationale, and appropriate documentation, for why this boundary/edge cases can occur and a rock-solid justification for why the handling given here is 100% Correct.

A hard and fast review rule in the Slang codebase should be: every single early-out `return`/`continue`/`break`/etc. should include a comment that fully justifies its existence. There is no reason why code like this should ever be accepted.

File: source/compiler-core/slang-rich-diagnostics-render.cpp [hunk 9]
```diff
@@ -915,10 +915,11 @@ String renderDiagnosticMachineReadable(
         // mirroring the logic in buildSectionLayout for rich diagnostics
         if (sll && span.range.begin == span.range.end && beginLoc.line > 0)
         {
-            SourceView* view = sm->findSourceView(span.range.begin);
+            SourceLoc spanBegin = span.range.begin;
+            SourceView* view = sm->findSourceViewThroughExpansion(spanBegin);
```
Comment (tangent-vector) [inline source/compiler-core/slang-rich-diagnostics-render.cpp:922]: Changes like this leave me wondering if this PR should actually be changing things so that `findSourceView` does the new thing by default, and we have a separate routine for the case where somebody *doesn't* want the expansion chain.

File: source/compiler-core/slang-source-loc.cpp [hunk 10]
```diff
@@ -803,6 +804,49 @@ SourceRange SourceManager::allocateSourceRange(UInt size)
...
+    {
+        if (manager->getSourceRange().contains(loc) && manager->m_macroExpansions.getCount() > 0)
+        {
+            const auto& entries = manager->m_macroExpansions;
+            // Binary search: entries are ordered by range.begin (monotone from
+            // allocateSourceRange).
+            Index lo = 0, hi = entries.getCount();
+            while (lo + 1 < hi)
+            {
+                Index mid = (lo + hi) >> 1;
```
Comment (tangent-vector) [inline source/compiler-core/slang-source-loc.cpp]: Boring comment that technically this approach can overflow and cause undefined behavior, so that it's really "supposed" to be `lo + (hi - lo)/2`.

File: source/compiler-core/slang-source-loc.cpp [hunk 11]
```diff
@@ -1019,17 +1063,29 @@ void SourceManager::addSourceFileIfNotExist(const String& uniqueIdentity, Source
     m_sourceFileMap.addIfNotExists(uniqueIdentity, sourceFile);
 }
 
+SourceView* SourceManager::findSourceViewThroughExpansion(SourceLoc& loc) const
+{
+    SourceView* view = findSourceViewRecursively(loc);
+    for (int depth = 0; depth < 64 && !view; ++depth)
```
Comment (tangent-vector) [inline source/compiler-core/slang-source-loc.cpp]: Magic numbers should become named constants. Honestly, stuff like these should probably eventually be configurable via compiler options, but leaving them hard-coded for now is fine.

File: source/compiler-core/slang-source-loc.cpp [hunk 12]
```diff
@@ -1019,17 +1063,29 @@ void SourceManager::addSourceFileIfNotExist(const String& uniqueIdentity, Source
...
+    {
+        const MacroExpansionEntry* entry = findMacroExpansion(loc);
+        if (!entry)
+            break;
+        loc = SourceLoc::fromRaw(
+            entry->bodyBegin.getRaw() + (loc.getRaw() - entry->range.begin.getRaw()));
+        view = findSourceViewRecursively(loc);
+    }
+    return view;
+}
```
Comment (tangent-vector) [inline source/compiler-core/slang-source-loc.cpp:1126]: This is a clear example of a subroutine that needs to start with a comment clearly stating:

- the problem to be solved
- the solution strategy that will be employed
- any relevant rationale for why that solution strategy is what it is, and what alternatives were considered but dismissed (and why)

As it is, this is the kind of subroutine I greatly dislike because its a "fun" little puzzle for the reader to work backwards from the implementation to the intent of the author, instead of being proper code that explains itself with intention, specifically so as *not* to puzzle a reader.

File: source/compiler-core/slang-source-loc.h [hunk 13]
```diff
@@ -577,6 +576,39 @@ struct SourceManager
...
+    /// renderer can emit "expanded from macro 'X'" notes without creating a SourceFile or
+    /// SourceView per invocation.
+    struct MacroExpansionEntry
+    {
+        SourceRange range;     ///< Tiny range of remapped source locs for this invocation's body.
+        String macroName;      ///< Macro name, for "expanded from macro 'X'" note.
+        SourceLoc callSiteLoc; ///< Where this macro was invoked (may itself be in an outer range).
+        SourceLoc bodyBegin;   ///< Original loc of the first body token; allows recovering the
+                               ///< original loc from a remapped one:
+                               ///<   originalLoc = bodyBegin + (remappedLoc - range.begin)
```
Comment (tangent-vector) [inline source/compiler-core/slang-source-loc.h]: The naming of this field isn't entirely clear about whether this is the location of the body of the macro *definition* or the *expansion*. The documentation comment makes it more clear, but at that point it seems like the correct name for the field is `originalBodyBeginLoc`.

File: source/compiler-core/slang-source-loc.h [hunk 14]
```diff
@@ -577,6 +576,39 @@ struct SourceManager
...
+    };
+
+    /// Register one macro invocation. Allocates a tiny source range of bodyRangeSize bytes,
+    /// appends the entry to m_macroExpansions, and returns the allocated range. The preprocessor
+    /// remaps body token locs into this range so each invocation's tokens are distinguishable.
+    /// bodyBegin is the original loc of the first body token.
```
Comment (tangent-vector) [inline source/compiler-core/slang-source-loc.h]: This "documentation" comment is describing how the function is *implemented* and not what it *does*, in terms that matter to somebody who calls it. A documentation comment should state the behavioral contract of a subroutine/type/whatever in terms that the client of the abstraction cares about. Implementation strategy belongs in implementation comments.

File: source/slang/slang-preprocessor.cpp [hunk 15]
```diff
@@ -779,6 +779,24 @@ struct MacroInvocation : InputStream
     /// nested macro invocations might be in flight.
     SourceLoc m_initiatingMacroInvocationLoc;
 
+    /// Tiny source range allocated in the SourceManager side table for this invocation's body
+    /// tokens. begin.isValid() iff this invocation is being tracked for diagnostic backtrace.
+    SourceRange m_expansionRange;
```
Comment (tangent-vector) [inline source/slang/slang-preprocessor.cpp]: The recurring use of "tiny" in comments related to this feature is a little offputting for a reader. It comes across more as advertisement rather than a technically-precise explanation of what's going on.

**Review by tangent-vector — CHANGES_REQUESTED** [review]
The main thing this PR needs now is an overhaul on the clarity of presentation to ensure that a future maintainer can properly read the code and understand what's going on. The clarity review skills checked into the Slang repository might be able to help with that.
